sdks/python/apache_beam/io/gcp/bigquery_read_internal.py

```diff
@@ -100,3 +127,264 @@ def process(self, unused_element, unused_signal, gcs_locations):
     )
 
     return main_output
+
+
+class _BigQueryReadSplit(beam.transforms.DoFn):
+  """Starts the process of reading from BigQuery.
+
+  This transform will start a BigQuery export job, and output a number of
+  file sources that are consumed downstream.
+  """
+  def __init__(
+      self,
+      options: PipelineOptions,
+      gcs_location: Union[str, ValueProvider] = None,
+      use_json_exports: bool = False,
+      bigquery_job_labels: Dict[str, str] = None,
+      step_name: str = None,
+      job_name: str = None,
+      unique_id: str = None,
+      kms_key: str = None,
+      project: str = None,
+      temp_dataset: Union[str, DatasetReference] = None):
+    self.options = options
+    self.use_json_exports = use_json_exports
+    self.gcs_location = gcs_location
+    self.bigquery_job_labels = bigquery_job_labels or {}
+    self._step_name = step_name
+    self._job_name = job_name or 'BQ_READ_SPLIT'
+    self._source_uuid = unique_id
+    self.kms_key = kms_key
+    self.project = project
+    self.temp_dataset = temp_dataset or 'bq_read_all_%s' % uuid.uuid4().hex
+    self.bq_io_metadata = None
+
+  def display_data(self):
+    return {
+        'use_json_exports': str(self.use_json_exports),
+        'gcs_location': str(self.gcs_location),
+        'bigquery_job_labels': json.dumps(self.bigquery_job_labels),
+        'kms_key': str(self.kms_key),
+        'project': str(self.project),
+        'temp_dataset': str(self.temp_dataset)
+    }
+
+  def _get_temp_dataset(self):
+    if isinstance(self.temp_dataset, str):
+      return DatasetReference(
+          datasetId=self.temp_dataset, projectId=self._get_project())
+    else:
+      return self.temp_dataset
+
+  def process(self, element: 'ReadFromBigQueryRequest', *args,
+              **kwargs) -> Iterable[BoundedSource]:
+    bq = bigquery_tools.BigQueryWrapper(
+        temp_dataset_id=self._get_temp_dataset().datasetId)
+
+    if element.query is not None:
+      self._setup_temporary_dataset(bq, element)
+      table_reference = self._execute_query(bq, element)
+    else:
+      assert element.table
+      table_reference = bigquery_tools.parse_table_reference(
+          element.table, project=self._get_project())
+
+    if not table_reference.projectId:
+      table_reference.projectId = self._get_project()
+
+    schema, metadata_list = self._export_files(bq, element, table_reference)
+
+    for metadata in metadata_list:
+      yield self._create_source(metadata.path, schema)
+
+    if element.query is not None:
+      bq._delete_table(
```

Comment:
We should also cleanup the temporary datasets that are created by the pipeline (may be at the end of every window).